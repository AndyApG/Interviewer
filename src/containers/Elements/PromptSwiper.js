import React, { Component } from 'react';
import { connect } from 'react-redux';
import Touch from 'react-hammerjs';
import PropTypes from 'prop-types';
import { findIndex } from 'lodash';
import { Prompt, Pips } from '../../components/Elements';

/**
  * Displays a control to swipe through prompts
  * @extends Component
  */
class PromptSwiper extends Component {
  constructor(props) {
    super(props);

    this.handleTap = this.handleTap.bind(this);
    this.handleSwipe = this.handleSwipe.bind(this);
  }

  handleSwipe(event) {
    switch (event.direction) {
      case 2:
      case 3:
        this.props.forward();
        break;
      case 1:
      case 4:
        this.props.backward();
        break;
      default:
    }
  }

  handleTap() {
    this.props.forward();
  }

  render() {
    const {
      promptIndex,
      prompts,
    } = this.props;

    const promptsRender = prompts.map((prompt, index) =>
      <Prompt key={index} label={prompt.title} isActive={promptIndex === index} />,
    );

    const classes = cx(
      'prompts',
      { 'prompts--floating': this.props.floating },
    );

    return (
      <Touch onTap={this.handleTap} onSwipe={this.handleSwipe} >
        <div className={classes}>
          <div className="prompts__pips">
            <Pips count={prompts.length} currentIndex={promptIndex} />
          </div>

          <div className="prompts__prompts">
            {promptsRender}
          </div>
        </div>
      </Touch>
    );
  }
}

PromptSwiper.propTypes = {
  forward: PropTypes.func.isRequired,
  backward: PropTypes.func.isRequired,
  prompts: PropTypes.any.isRequired,
  promptIndex: PropTypes.number.isRequired,
  floating: PropTypes.bool,
};

PromptSwiper.defaultProps = {
  floating: false,
};

function mapStateToProps(state, ownProps) {
  return {
    promptIndex: findIndex(ownProps.prompts, ownProps.prompt),
  };
}

export default connect(mapStateToProps)(PromptSwiper);
