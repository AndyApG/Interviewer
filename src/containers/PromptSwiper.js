import React, { Component } from 'react';
import { connect } from 'react-redux';
import Touch from 'react-hammerjs';
import PropTypes from 'prop-types';
import { findIndex } from 'lodash';
import cx from 'classnames';
import { Prompt, Pips } from '../components/';

/**
  * Displays a control to swipe through prompts
  * @extends Component
  */
class PromptSwiper extends Component {
  static propTypes = {
    forward: PropTypes.func.isRequired,
    backward: PropTypes.func.isRequired,
    prompts: PropTypes.any.isRequired,
    promptIndex: PropTypes.number.isRequired,
    floating: PropTypes.bool,
    minimizable: PropTypes.bool,
  };

  static defaultProps = {
    floating: false,
    minimizable: false,
  };

  constructor(props) {
    super(props);

    this.state = {
      minimized: false,
    };
    this.handleTap = this.handleTap.bind(this);
    this.handleSwipe = this.handleSwipe.bind(this);
  }

  handleSwipe(event) {
    switch (event.direction) {
      case 2:
      case 3:
        this.props.forward();
        break;
      case 1:
      case 4:
        this.props.backward();
        break;
      default:
    }
  }

  handleTap() {
    this.props.forward();
  }

  handleMinimize = () => {
    this.setState({
      minimized: !this.state.minimized,
    });
  };

  render() {
    const {
      minimizable,
      promptIndex,
      prompts,
    } = this.props;

    const promptsRender = prompts.map((prompt, index) =>
      (<Prompt
        key={index}
        label={prompt.text}
        isActive={promptIndex === index}
        isLeaving={promptIndex === (index - 1)}
      />),
    );

    const classes = cx(
      'prompts',
      { 'prompts--floating': this.props.floating,
        'prompts--minimized': this.state.minimized },
    );

    if (prompts.length <= 1) {
      return (
        <div className={classes}>
          <div className="prompts__prompts">
            {promptsRender}
          </div>
        </div>
      );
    }

    return (
      <React.Fragment>
        <Touch onTap={this.handleTap} onSwipe={this.handleSwipe} >
          <div className={classes}>
            <div className="prompts__pips">
              <Pips count={prompts.length} currentIndex={promptIndex} />
            </div>
            {!this.state.minimized && (<div className="prompts__prompts">
              {promptsRender}
            </div>)}
          </div>
        </Touch>
        {minimizable &&
          <span className="prompts__minimizer" onClick={this.handleMinimize}>
            {this.state.minimized ? '?' : 'â€”'}
          </span>
        }
      </React.Fragment>
    );
  }
}

function mapStateToProps(state, ownProps) {
  return {
    promptIndex: findIndex(ownProps.prompts, ownProps.prompt),
  };
}

export { PromptSwiper };

export default connect(mapStateToProps)(PromptSwiper);
