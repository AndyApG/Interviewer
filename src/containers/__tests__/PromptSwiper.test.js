/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import { PromptSwiper } from '../PromptSwiper';

jest.mock('../../utils/CSSVariables');

const mockProps = {
  forward: () => {},
  backward: () => {},
  prompts: [
    { text: 'bar' },
    { text: 'foo' },
  ],
  promptIndex: 0,
};

describe('<PromptSwiper />', () => {
  it('renders ok', () => {
    const component = shallow(<PromptSwiper {...mockProps} />);

    expect(component.find('.prompts__pips').exists()).toBe(true);
    expect(component).toMatchSnapshot();
  });
});

describe("when only one prompt, don't show pips", () => {
  it('renders ok', () => {
    const prompts = [{ text: 'baz' }];
    const component = shallow(<PromptSwiper {...mockProps} prompts={prompts} />);

    expect(component.find('.prompts__pips').exists()).toBe(false);
  });
});
