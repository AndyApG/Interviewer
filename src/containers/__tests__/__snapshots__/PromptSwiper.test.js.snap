// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PromptSwiper /> renders ok 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PromptSwiper
    backward={[Function]}
    floating={false}
    forward={[Function]}
    minimizable={false}
    promptIndex={0}
    prompts={
      Array [
        Object {
          "text": "bar",
        },
        Object {
          "text": "foo",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <div
          className="prompts"
        >
          <div
            className="prompts__pips"
          >
            <Pips
              count={2}
              currentIndex={0}
              large={false}
            />
          </div>
          <div
            className="prompts__prompts"
          >
            <Prompt
              isActive={true}
              isLeaving={false}
              label="bar"
            />
            <Prompt
              isActive={false}
              isLeaving={true}
              label="foo"
            />
          </div>
        </div>,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="prompts__pips"
            >
              <Pips
                count={2}
                currentIndex={0}
                large={false}
              />
            </div>,
            <div
              className="prompts__prompts"
            >
              <Prompt
                isActive={true}
                isLeaving={false}
                label="bar"
              />
              <Prompt
                isActive={false}
                isLeaving={true}
                label="foo"
              />
            </div>,
          ],
          "className": "prompts",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Pips
                count={2}
                currentIndex={0}
                large={false}
              />,
              "className": "prompts__pips",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "count": 2,
                "currentIndex": 0,
                "large": false,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Prompt
                  isActive={true}
                  isLeaving={false}
                  label="bar"
                />,
                <Prompt
                  isActive={false}
                  isLeaving={true}
                  label="foo"
                />,
              ],
              "className": "prompts__prompts",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "0",
                "nodeType": "function",
                "props": Object {
                  "isActive": true,
                  "isLeaving": false,
                  "label": "bar",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "1",
                "nodeType": "function",
                "props": Object {
                  "isActive": false,
                  "isLeaving": true,
                  "label": "foo",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
      false,
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <div
            className="prompts"
          >
            <div
              className="prompts__pips"
            >
              <Pips
                count={2}
                currentIndex={0}
                large={false}
              />
            </div>
            <div
              className="prompts__prompts"
            >
              <Prompt
                isActive={true}
                isLeaving={false}
                label="bar"
              />
              <Prompt
                isActive={false}
                isLeaving={true}
                label="foo"
              />
            </div>
          </div>,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="prompts__pips"
              >
                <Pips
                  count={2}
                  currentIndex={0}
                  large={false}
                />
              </div>,
              <div
                className="prompts__prompts"
              >
                <Prompt
                  isActive={true}
                  isLeaving={false}
                  label="bar"
                />
                <Prompt
                  isActive={false}
                  isLeaving={true}
                  label="foo"
                />
              </div>,
            ],
            "className": "prompts",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Pips
                  count={2}
                  currentIndex={0}
                  large={false}
                />,
                "className": "prompts__pips",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "count": 2,
                  "currentIndex": 0,
                  "large": false,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Prompt
                    isActive={true}
                    isLeaving={false}
                    label="bar"
                  />,
                  <Prompt
                    isActive={false}
                    isLeaving={true}
                    label="foo"
                  />,
                ],
                "className": "prompts__prompts",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "0",
                  "nodeType": "function",
                  "props": Object {
                    "isActive": true,
                    "isLeaving": false,
                    "label": "bar",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": "1",
                  "nodeType": "function",
                  "props": Object {
                    "isActive": false,
                    "isLeaving": true,
                    "label": "foo",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
        false,
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
